simapp/ante.go

```diff
@@ -0,0 +1,33 @@
+package simapp
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth"
+	authante "github.com/cosmos/cosmos-sdk/x/auth/ante"
+	"github.com/cosmos/cosmos-sdk/x/feegrant"
+	feegrantante "github.com/cosmos/cosmos-sdk/x/feegrant/ante"
+	feegrantexported "github.com/cosmos/cosmos-sdk/x/feegrant/exported"
+)
+
+// NewAnteHandler returns an AnteHandler that checks and increments sequence
+// numbers, checks signatures & account numbers, and deducts fees from the first
+// signer.
+func NewAnteHandler(
+	ak auth.AccountKeeper, supplyKeeper feegrantexported.SupplyKeeper, feeGrantKeeper feegrant.Keeper,
+	sigGasConsumer auth.SignatureVerificationGasConsumer,
+) sdk.AnteHandler {
+
+	return sdk.ChainAnteDecorators(
```

Comment:
@aaronc @ethanfrey please verify the ante-handler and order here.

Reply:
It has been a while, but this looks good.

`NewSetPubKeyDecorator` position is key, as otherwise an empty account could never use a grant until it had some tokens (which is not the behavior we wanted)

Reply:
Starting to take a look at this work again as I'm trying to figure out how to deal with account numbers and sequences starting from 1 (which was a part of ADR 020).

So one thing that I'm aware of here is that if we create the account in the `DeduceGrantedFeeDecorator` that account number will differ from the one use for tx signing. So it seems like the correct approach to doing this is to allow accounts that don't yet exist in state to use some zero-value for account number to do signing. Since we want to avoid `0` as a value for account number, it sounds like `1` would become the zero value and `2` would be the first number of any real account in state.

What do you think?

Reply:
Another alternative is that accounts get created (if they don't exist) whenever fee grants are created. That might be a bit cleaner in some ways.

Although, conceptually the idea that there is _some_ way for accounts that aren't in state to sign transactions sounds desirable to me.